ext/odbc/php_odbc.c

```diff
@@ -65,59 +65,49 @@
 
 void odbc_do_connect(INTERNAL_FUNCTION_PARAMETERS, int persistent);
 static void safe_odbc_disconnect(void *handle);
-static void close_results_with_connection(const odbc_connection *conn);
+static void close_results_with_connection(odbc_connection *conn);
+static inline odbc_result *odbc_result_from_obj(zend_object *obj);
 
 static int le_pconn;
 
 static zend_class_entry *odbc_connection_ce, *odbc_result_ce;
 static zend_object_handlers odbc_connection_object_handlers, odbc_result_object_handlers;
 
-static void odbc_insert_new_result(zval *result) {
-	Z_TRY_ADDREF_P(result);
-	zend_hash_next_index_insert_new(&ODBCG(results), result);
-}
-
-static inline odbc_link *odbc_link_from_obj(zend_object *obj) {
-	return (odbc_link *)((char *)(obj) - XtOffsetOf(odbc_link, std));
-}
-
 #define Z_ODBC_LINK_P(zv) odbc_link_from_obj(Z_OBJ_P(zv))
 #define Z_ODBC_CONNECTION_P(zv) Z_ODBC_LINK_P(zv)->connection
+#define Z_ODBC_RESULT_P(zv) odbc_result_from_obj(Z_OBJ_P(zv))
 
-static zend_object *odbc_connection_create_object(zend_class_entry *class_type) {
-	odbc_link *intern = zend_object_alloc(sizeof(odbc_link), class_type);
-
-	zend_object_std_init(&intern->std, class_type);
-	object_properties_init(&intern->std, class_type);
-	intern->std.handlers = &odbc_connection_object_handlers;
-
-	return &intern->std;
-}
+static void odbc_insert_new_result(odbc_connection *connection, zval *result) {
+	ZEND_ASSERT(Z_TYPE_P(result) == IS_OBJECT && instanceof_function(Z_OBJCE_P(result), odbc_result_ce));
+	if (!connection->results) {
+		connection->results = zend_new_array(1);
+	}
 
-static zend_function *odbc_connection_get_constructor(zend_object *object) {
-	zend_throw_error(NULL, "Cannot directly construct ODBC\\Connection, use odbc_connect() or odbc_pconnect() instead");
-	return NULL;
+	zend_hash_next_index_insert_new(connection->results, result);
+	odbc_result *res = Z_ODBC_RESULT_P(result);
+	res->index = zend_hash_num_elements(connection->results) - 1;
+	Z_ADDREF_P(result);
 }
 
 static void odbc_link_free(odbc_link *link)
 {
 	ZEND_ASSERT(link->connection && "link has already been closed");
 
-	if (link->persistent) {
-		zend_hash_del(&EG(persistent_list), link->hash);
```

Comment:
I realized that the persistent list only stores the connections themselves so these should only be freed when a connection/all connections are explicitly closed, or during module shutdown. AFAIR I still need to implement freeing persistent connections for odbc_close(). It's done for odbc_close_all() though.